wallet/src/de/schildbach/wallet/service/CoinJoinService.kt

```diff
@@ -0,0 +1,323 @@
+    }
+
+    private val coinJoinManager: CoinJoinManager?
+        get() = walletDataProvider.wallet?.context?.coinJoinManager
+    private lateinit var clientManager: CoinJoinClientManager
+
+    private var mixingCompleteListeners: ArrayList<MixingCompleteListener> = arrayListOf()
+    private var sessionCompleteListeners: ArrayList<SessionCompleteListener> = arrayListOf()
+
+    private var mixingStatus: MixingStatus = MixingStatus.NOT_STARTED
+
+    private val syncScope = CoroutineScope(
+        Executors.newFixedThreadPool(5).asCoroutineDispatcher()
+    )
+
+    private var blockChain: AbstractBlockChain? = null
+    private var isBlockChainSet = false
+    private var networkStatus: NetworkStatus = NetworkStatus.STOPPED
+
+    // https://stackoverflow.com/questions/55421710/how-to-suspend-kotlin-coroutine-until-notified            //mixingStatus = MixingStatus.MIXING
+
+    private val mixingMutex = Mutex(locked = true)
+    private var exception: Throwable? = null
+    override suspend fun waitForMixing() = mixingMutex.withLock{}
+    override suspend fun waitForMixingWithException() {
+        waitForMixing()
+        exception?.let {
+            throw it
+        }
+    }
+    private fun setMixingComplete() {
+        mixingMutex.unlock()
+    }
+
+    init {
+        blockchainStateProvider.observeNetworkStatus()
+            .onEach { status ->
+                updateNetworkStatus(status)
+            }
+            .launchIn(syncScope)
+
+        blockchainStateProvider.observeBlockChain()
+            .filterNotNull()
+            .onEach { blockChain ->
+                updateBlockChain(blockChain)
+            }
+            .launchIn(syncScope)
+    }
+
+    private suspend fun updateState(mixingStatus: MixingStatus, networkStatus: NetworkStatus, blockChain: AbstractBlockChain?) {
+
+        log.info("coinjoin-updateState: $mixingStatus, $networkStatus, ${blockChain != null}")
+        setBlockchain(blockChain)
+
+        when {
+            mixingStatus == MixingStatus.MIXING && networkStatus == NetworkStatus.CONNECTED -> {
+                prepareAndStartMixing2()
+                setBlockchain(blockChain)
+                startMixing()
+            }
+            this.mixingStatus == MixingStatus.MIXING && mixingStatus == MixingStatus.FINISHED -> {
+                stopMixing()
+                setMixingComplete()
+            }
+            mixingStatus == MixingStatus.MIXING && this.networkStatus == NetworkStatus.CONNECTED && networkStatus == NetworkStatus.NOT_AVAILABLE -> {
+                stopMixing()
+            }
+            mixingStatus == MixingStatus.PAUSED && this.networkStatus == NetworkStatus.NOT_AVAILABLE && networkStatus == NetworkStatus.CONNECTED -> {
+                startMixing()
+            }
+            mixingStatus == MixingStatus.ERROR -> setMixingComplete()
+        }
+        this.networkStatus = networkStatus
+        this.mixingStatus = mixingStatus
+    }
```

Comment:
This is the function that will start and stop mixing based on various inputs such as what do we want to do (mix) and whether are we connected to the Dash Network.

Reply:
Not fully tested yet.